packages/connect-query/src/connect-query-key.ts

```diff
@@ -12,73 +12,181 @@
+      schema: Desc;
+      /**
+       * Set `transport` in the key.
+       */
+      transport?: Transport;
+      /**
+       * Set `cardinality` in the key - "finite" by default.
+       */
+      cardinality?: "finite" | "infinite" | "any";
+    };
+
 /**
- * TanStack Query requires query keys in order to decide when the query should automatically update.
+ * TanStack Query manages query caching for you based on query keys. In Connect Query, keys are structured, and can easily be created using this factory function.
+ *
+ * When you make a query, a unique key is automatically created from the schema, input message, and transport. For example:
+ *
+ * ```ts
+ * import { useQuery } from "@connectrpc/connect-query";
+ *
+ * useQuery(ElizaService.method.say, { sentence: "hello" });
+ *
+ * // creates the key:
+ * [
+ *   "connect-query",
+ *   {
+ *     transport: "t1",
+ *     serviceName: "connectrpc.eliza.v1.ElizaService",
+ *     methodName: "Say",
+ *     input: { sentence: "hello" },
+ *     cardinality: "finite",
+ *   }
+ * ]
+ * ```
+ *
+ * The same key can be created manually with this factory:
  *
- * In Connect-Query, much of this is handled automatically by this function.
+ * ```ts
+ * createConnectQueryKey({
+ *   transport: myTransportReference,
+ *   schema: ElizaService.method.say,
+ *   input: { sentence: "hello" }
+ * });
+ * ```
+ *
+ * Note that the factory allows to create partial keys that can be used to filter queries. For example, you can create a key without a transport, any cardinality, any input message, or with a partial input message.
  *
  * @see ConnectQueryKey for information on the components of Connect-Query's keys.
  */
-export function createConnectQueryKey<
-  I extends DescMessage,
-  O extends DescMessage,
->(
-  schema: Pick<MethodUnaryDescriptor<I, O>, "input" | "parent" | "name">,
-  input?: SkipToken | MessageInitShape<I> | undefined,
+export function createConnectQueryKey<Desc extends DescMethod | DescService>(
+  params: KeyParams<Desc>,
 ): ConnectQueryKey {
-  const key =
-    input === skipToken || input === undefined
-      ? createMessageKey(schema.input, {} as MessageInitShape<DescMessage & I>)
-      : createMessageKey(schema.input, input);
-  return [schema.parent.typeName, schema.name, key];
-}
-
-/**
- * Similar to @see ConnectQueryKey, but for infinite queries.
- */
-export type ConnectInfiniteQueryKey = [
-  serviceTypeName: string,
-  methodName: string,
-  input: Record<string, unknown>,
-  "infinite",
-];
-
-/**
- * Similar to @see createConnectQueryKey, but for infinite queries.
- */
-export function createConnectInfiniteQueryKey<
-  I extends DescMessage,
-  O extends DescMessage,
->(
-  schema: Pick<MethodUnaryDescriptor<I, O>, "input" | "parent" | "name">,
-  input?: SkipToken | MessageInitShape<I> | undefined,
-): ConnectInfiniteQueryKey {
-  return [...createConnectQueryKey(schema, input), "infinite"];
+  const props: ConnectQueryKey[1] =
+    params.schema.kind == "rpc"
```

Comment:
Hmm, interesting. I guess this allows for invalidating a whole service if necessary.